play-frontend-hmrc-play-30/src/main/twirl/uk/gov/hmrc/govukfrontend/views/components/GovukServiceNavigation.scala.html

```diff
@@ -22,68 +22,69 @@
+        @serviceUrl match {
+          case Some(url) => {
+            <a href="@url" class="govuk-service-navigation__link"> @serviceName </a>
+            }
+          case None => {
+            <span class="govuk-service-navigation__text">@{serviceName.get}</span>
+          }
+        }
+      </span>
+    }
+    @if(navigation.nonEmpty || slots.exists(slot => slot.navigationStart.nonEmpty || slot.navigationEnd.nonEmpty)) {
+      <nav aria-label="@if(navigationLabel.nonEmpty){@navigationLabel}else{@menuButtonText.getOrElse(messages("govuk.serviceNavigation.menuButtonText"))}" class="@{toClasses("govuk-service-navigation__wrapper", navigationClasses)}">
+        <button type="button" class="govuk-service-navigation__toggle govuk-js-service-navigation-toggle" aria-controls="@navigationId"@if(menuButtonLabel.nonEmpty && menuButtonLabel.get != menuButtonText.getOrElse(messages("govuk.serviceNavigation.menuButtonText"))){ aria-label="@menuButtonLabel"} hidden>
+          @menuButtonText.getOrElse(messages("govuk.serviceNavigation.menuButtonText"))
+        </button>
+        <ul class="govuk-service-navigation__list" id="@navigationId">
+          @Html(slots.flatMap(_.navigationStart))
+          @for(item <- navigation) {
+            <li class="govuk-service-navigation__item@if(item.active || item.current){ govuk-service-navigation__item--active}">
+              @if(item.href.nonEmpty) {
+                <a @if(item.active || item.current){aria-current="@if(item.current){page}else{true}"} class="govuk-service-navigation__link" href="@item.href" @toAttributes(item.attributes)>
+              }else{
+                <span class="govuk-service-navigation__text">
+              }
+              @if(item.current || item.active) {
+                <strong class="govuk-service-navigation__active-fallback">@item.content.asHtml</strong>
+              } else {
+                @item.content.asHtml
+              }
+              @if(item.href.nonEmpty) {
+                </a>
+              } else {
+                </span>
+              }
+            </li>
+          }
+          @Html(slots.flatMap(_.navigationEnd))
+        </ul>
+      </nav>
+    }
+  </div>
+}
+
+@govukWidthContainer = {
+  <div class="govuk-width-container">
+    @Html(slots.flatMap(_.start))
+
+    @govukServiceNavigationContainer
+
+    @Html(slots.flatMap(_.end))
+  </div>
+}
+
 @if(params.serviceName.isDefined || params.slots.flatMap(_.start).isDefined || params.slots.flatMap(_.end).isDefined) {
   <section aria-label="@{ariaLabel.getOrElse(messages("govuk.serviceNavigation.ariaLabel"))}" class="@toClasses("govuk-service-navigation", classes)" data-module="govuk-service-navigation" @toAttributes(attributes)>
+    @govukWidthContainer
+  </section>
 } else {
   <div class="@toClasses("govuk-service-navigation", classes)" data-module="govuk-service-navigation" @toAttributes(attributes)>
+    @govukWidthContainer
+  </div>
```

Comment:
it's easy to miss balancing tags when you have multiple if statements that need to change together, so rather than that, I've extracted some variables - now we could get help from IDE too if stuff is unbalanced (though won't pick up everything)

Reply:
you can see the failing test cases for this if you look at the build of the last commit, where I added tests to check for unbalanced html

Reply:
I wiped had to rebase on main, so I've lost the build - but here was the test result showing it failing on unbalanced tags in service navigation

![CleanShot 2024-11-04 at 16 10 10@2x](https://github.com/user-attachments/assets/224503c9-eb8f-4f31-9f28-36e6a3589923)